fabtests/component/sock_test.c

```diff
@@ -0,0 +1,253 @@
+		if (ret)
+			return ret;
+
+		(*clients)[i] = new_sock;
+	}
+
+	return ret;
+}
+
+double time_select(int connections, int server_sock, int *clients) 
+{
+	fd_set select_set;
+	int i, max_sock;
+	struct timeval timeout;
+	struct timespec start;
+	struct timespec stop;
+
+	timeout.tv_sec = 0;
+	timeout.tv_usec = 0;
+
+	FD_ZERO(&select_set);
+	FD_SET(server_sock, &select_set);
+
+	max_sock = server_sock;
+
+	for (i = 0; i < connections; i++) {
+		FD_SET(clients[i], &select_set);
+
+		if (clients[i] > max_sock)
+			max_sock = clients[i];
+
+	}
+
+	clock_gettime(CLOCK_MONOTONIC, &start);
+	select(max_sock + 1, &select_set, NULL, NULL, &timeout);
+	clock_gettime(CLOCK_MONOTONIC, &stop);
+
+	return (stop.tv_nsec + stop.tv_sec * 1000000000) - (start.tv_nsec + start.tv_sec * 1000000000);
+}
+
+double time_poll(int connections, int server_sock, int *clients)
+{
+	struct pollfd *poll_set;
+	int i;
+	struct timespec start;
+	struct timespec stop;
+
+	poll_set = malloc(sizeof(struct pollfd) * connections);
+
+	for (i = 0; i < connections; i++) {
+		poll_set[i].fd = clients[i];
+	}
+
+	clock_gettime(CLOCK_MONOTONIC, &start);
+	poll(poll_set, POLLIN, 0);
+	clock_gettime(CLOCK_MONOTONIC, &stop);
+
+	return (stop.tv_nsec + stop.tv_sec * 1000000000) - (start.tv_nsec + start.tv_sec * 1000000000);
+}
+
+double time_epoll(int connections, int server_sock, int *clients)
+{
+	struct epoll_event *events;
+	int efpd = epoll_create1(0);
+	int i, ret;
+	struct timespec start;
+	struct timespec stop;
+
+	events = malloc(sizeof(struct epoll_event) * connections);
+
+	for (i = 0; i < connections; i++) {
+
+		events[i].events = EPOLLIN;
+    	events[i].data.fd = clients[i];
+
+		ret = epoll_ctl(efpd, EPOLL_CTL_ADD, clients[i], &events[i]);
+		if (ret)
+			return ret;
+	}
+
+	clock_gettime(CLOCK_MONOTONIC, &start);
+	epoll_wait(EPOLLIN, events, connections, 0);
+	clock_gettime(CLOCK_MONOTONIC, &stop);
+	return (stop.tv_nsec + stop.tv_sec * 1000000000) - (start.tv_nsec + start.tv_sec * 1000000000);
+
+}	
+
+int main(int argc, char **argv)
+{
+	extern char *optarg;
+	int c, ret, connections, sock, i;
+	char *server_address, *mode;
+	struct sockaddr_storage server_addr;
+	int *clients;
+	double elapsed;
+
+	while ((c = getopt(argc, argv, "n:s:m:")) != -1) {
+		switch (c) {
+		case 's':
+			server_address = optarg;
+			break;
+		case 'n':
+			connections = atoi(optarg);
+			break;
+		case 'm':
+			mode = optarg;
+			break;
+		}
+	}
+
+	ret = get_server_addr(server_address, &server_addr, &sock);
+	if (ret == 0) {
+		ret = server_create(connections, server_addr, sock, &clients);
+		if (ret)
+			return ret;
+
+		if (strcmp(mode, "select") == 0) {
+			printf("select mode\n");
+			elapsed = time_select(connections, sock, clients);
+		} else if (strcmp(mode, "poll") == 0) {
+			printf("poll mode\n");
+			elapsed = time_poll(connections, sock, clients);
+		} else {
+			printf("epoll mode\n");
+			elapsed = time_epoll(connections, sock, clients);

```

Comment:
We're going to need to split the above functions (time_select, etc.) into 2 components.  The first part is an initialization.  We could probably do that as part of some common init code.  For example both the client and server should call something like init_resources().  There allocate the socket array.  Also allocate other resources, such as the pollfd array, the epoll fd, etc.

The second part is to keep the part of the function that calls select/poll/epoll.  We'll want to call those functions in a loop a few thousand times, so that we can get an average of how long each function takes, versus relying on the timing of a single call.